programs/sig-aggregator/src/attestor_data.rs

```diff
@@ -16,76 +14,163 @@ type Signature = FixedBytes<SIGNATURE_BYTE_LENGTH>;
-type SignedStates = HashMap<State, Vec<(Signature, Pubkey)>>;
+/// Maps attested_data -> list of attestations
+///
+/// Structure:
+/// ```text
+/// Attested_data: 0x1234... (12 bytes)
+///     [Attestation_A, Attestation_B]
+/// Attested_data: 0x9876...
+///     [Attestation_C, Attestation_D]
+/// ```
 
-#[derive(Debug, Clone)]
-pub struct AttestatorData(HashMap<Height, SignedStates>);
+#[derive(Debug, Clone, Default)]
+pub struct AttestatorData {
+    state_attestations: HashMap<State, Vec<Attestation>>,
+}
 
 impl AttestatorData {
+    #[must_use]
     pub fn new() -> Self {
-        Self(HashMap::new())
+        Self::default()
     }
 
-    pub fn insert(&mut self, att_resp: AttestationsFromHeightResponse) {
-        for attestations in att_resp.attestations {
-            let state_map = self.0.entry(attestations.height).or_default();
-            state_map
-                .entry(State::from_slice(&attestations.data))
-                .or_default()
-                .push((
-                    Signature::from_slice(&attestations.signature),
-                    Pubkey::from_slice(&att_resp.pubkey),
-                ));
+    pub fn insert(&mut self, attestation: Attestation) {
+        if let Err(e) = attestation.validate() {
+            tracing::error!("Invalid attestation should not happen: {e}");
+            return;
         }
-    }
 
-    pub fn get_latest(&self, quorum: usize) -> Option<AggregateResponse> {
-        let mut latest = AggregateResponse {
-            height: 0,
-            state: vec![],
-            sig_pubkey_pairs: vec![],
-        };
+        let attested_data = State::try_from(attestation.attested_data.as_slice())
+            .expect("Failed to convert attested_data to State: invalid length or format");
+        if let Some(attestations) = self.state_attestations.get_mut(&attested_data) {
+            attestations.push(attestation);
+            return;
+        }
+        self.state_attestations
+            .insert(attested_data, vec![attestation]);
+    }
 
-        for (height, state_map) in self.0.iter() {
-            if *height <= latest.height {
-                continue;
+    #[must_use]
+    pub fn get_quorum(&self, quorum: usize) -> Option<AggregateResponse> {
+        for (state, attestations) in &self.state_attestations {
+            let count = attestations.len();
+            if count >= quorum {
+                return Some(AggregateResponse {
+                    height: attestations[0].height,
+                    state: state.to_vec(),
+                    sig_pubkey_pairs: attestations
+                        .iter()
+                        .map(|a| SigPubkeyPair {
+                            sig: a.signature.clone(),
+                            pubkey: a.public_key.clone(),
+                        })
+                        .collect(),
+                });
             }
+        }
+        None
```

Comment:
Seems like this could be done in a more ideomatic rust fashion with iterator instead of the a for loop.

Reply:
Something like
```rust
pub fn get_quorum(&self, quorum: usize) -> Option<AggregateResponse> {
    self.state_attestations
        .iter()
        .find(|(_, attestations)| attestations.len() >= quorum)
        .map(|(state, attestations)| AggregateResponse {
            height: attestations[0].height,
            state: state.to_vec(),
            sig_pubkey_pairs: attestations
                .iter()
                .map(|a| SigPubkeyPair {
                    sig: a.signature.clone(),
                    pubkey: a.public_key.clone(),
                })
                .collect(),
        })
}
```